### PR TITLE
added iterator to collection/index

### PR DESCRIPTION
**Pull request by woutslakhorst**
This allows for functionality on indices like UNIQUE, SIZE etc in a performant way.

### CODE REVIEW

**Comment by reinkrul** [comment]
I think I need some more in-depth explanation to understand this.

File: collection.go [hunk 1]
```diff
@@ -185,33 +192,60 @@ func (c *collection) Find(query Query) ([]Document, error) {
...
+	// the iteratorFn that collects results in a slice
+	refMap := map[string]bool{}
+	var newRef = make([]Reference, 0)
+	var refFn = func(key []byte, value []byte) error {
+		refs, err := entryToSlice(value)
 		if err != nil {
 			return err
 		}
+		for _, r := range refs {
+			if _, b := refMap[r.EncodeToString()]; !b {
```
Comment (reinkrul) [inline collection.go:208]: can everything that can be expressed in bytes be safely represented as string (without collisions)?

I think you can index on an array (only works if the lengths are fixed though)

**Review by reinkrul — APPROVED** [review]
small comment though

File: collection.go (hunk 1 above)
Comment (woutslakhorst) [reply collection.go:208]: they are `references`, a custom type which does hex encoding of byte arrays.

**Pull request merged**